syntax = "proto3";
package vmr_proto;

// import "github.com/gogo/protobuf/gogoproto/gogo.proto";
import "google/protobuf/empty.proto";

option go_package = "git.supremind.info/product/visionmind/proto/go/api";
// ref: https://github.com/gogo/protobuf/blob/master/extensions.md
// option (gogoproto.equal_all) = true;
// option (gogoproto.goproto_unrecognized_all) = false;
// option (gogoproto.populate_all) = true;


enum DeviceType {
    DEVICE_TYPE_HIKVISION   = 0;
    DEVICE_TYPE_STREAM_ADDR = 1;
    DEVICE_TYPE_GB28181     = 2;
    DEVICE_TYPE_ONVIF       = 3;
}

enum DeviceProtocolType {
     DEVICE_PROTOCOL_TYPE_DEFAULT = 0;
     DEVICE_PROTOCOL_TYPE_SDK = 1;      //  SDK
     DEVICE_PROTOCOL_TYPE_ONVIF = 3;    //  ONVIF
     DEVICE_PROTOCOL_TYPE_SIP = 4;      //  SIP
}

enum CameraType {
    CAMERA_TYPE_DEFAULT = 0;
    CAMERA_TYPE_DOME    = 1; //球机,支持ptz
    CAMERA_TYPE_GUN     = 3; //枪机等不支持ptz的ipc
}

message CreateStreamRequest {
    string device_id = 1;
    uint32 stream_rtp_port  = 2;
    string stream_host = 3;
}

message RetrieveStreamRequest {
    uint32 ssrc = 1;
}

message RetrieveStreamResponse {
    string device_id = 1;
}

message CreateStreamResponse {
    DeviceType  device_type    = 1;
    uint32      channel_id     = 2;
    string      host           = 3;
    string      user_name      = 4;
    string      password       = 5;
    string      stream_url     = 6;
    string      description    = 7;
    repeated string extra_url_list = 8;
    bool        mp4_loop_status = 9;
    int32       mp4_loop_count  = 10;
}

message DeleteStreamRequest {
    string device_id = 1;
}

message GetRecordListReq {
    uint32 record_id = 1;
}

message RecordListResponse {
    repeated string record_url = 1;
}

message GetDeviceListRequest {
    uint32 stream_index = 1;
}

message GetDeviceListResponse {
    repeated string device_id = 1;
}

message GetSubDeviceInfoRequest {
    string sub_device_id = 1;
}

message GetSubDeviceInfoResponse {
    string                  sub_device_name = 1;    // 子设备名称
    DeviceType              device_type = 2;        // 设备类型
    string                  device_id = 3;          // 主设备id
    int32                   channel_id = 4;         // 通道id
    CameraType              camera_type = 5;        // 相机类型
    DeviceProtocolType      protocol_type = 6;      // 协议类型
    string                  ip_address = 7;         // IP地址
    string                  account = 8;            // 账号
    string                  password = 9;           // 密码
    int32                   record_time_quota  = 10; // -1代表没有缓存时间字段，0代表不需要缓存，大于0代表有缓存时间
}

service FlowService {
    rpc CreateStream   (CreateStreamRequest) returns (CreateStreamResponse);
    rpc RetrieveStream (RetrieveStreamRequest) returns (RetrieveStreamResponse);
    rpc DeleteStream   (DeleteStreamRequest) returns (google.protobuf.Empty);
    rpc GetRecordList (GetRecordListReq) returns (RecordListResponse);
    rpc GetDeviceList (GetDeviceListRequest) returns (GetDeviceListResponse);
    rpc GetSubDeviceInfo (GetSubDeviceInfoRequest) returns (GetSubDeviceInfoResponse);
}